cpp/src/io/json/nested_json.hpp

```diff
@@ -185,6 +185,73 @@ struct device_json_column {
   }
 };
 
+namespace experimental {
+/*
+ * @brief Sparse graph adjacency matrix stored in Compressed Sparse Row (CSR) format.
+ */
+struct csr {
+  rmm::device_uvector<NodeIndexT> rowidx;
+  rmm::device_uvector<NodeIndexT> colidx;
+};
+
+/*
+ * @brief Auxiliary column tree properties that are required to construct the device json
+ * column subtree, but not required for the final cudf column construction.
+ */
+struct column_tree_properties {
+  rmm::device_uvector<NodeT> categories;
+  rmm::device_uvector<size_type> max_row_offsets;
+  rmm::device_uvector<NodeIndexT> mapped_ids;
+};
+
+/*
+ * @brief Unvalidated column tree stored in Compressed Sparse Row (CSR) format. The device json
+ * column subtree - the subgraph that conforms to column tree properties - is extracted and further
+ * processed according to the JSON reader options passed. Only the final processed subgraph is
+ * annotated with information required to construct cuDF columns.
+ */
+struct column_tree {
+  // position of nnzs
+  csr adjacency;
+  rmm::device_uvector<NodeIndexT> rowidx;
+  rmm::device_uvector<NodeIndexT> colidx;
+  // device_json_column properties
+  using row_offset_t = size_type;
+  // Indicator array for the device column subtree
+  // Stores the number of rows in the column if the node is part of device column subtree
+  // Stores zero otherwise
+  rmm::device_uvector<row_offset_t> subtree_nrows;
+  rmm::device_uvector<row_offset_t> string_offsets;
+  rmm::device_uvector<row_offset_t> string_lengths;
+  // Row offsets
+  rmm::device_uvector<row_offset_t> child_offsets;
+  // Validity bitmap
+  rmm::device_buffer validity;
+};
+
+namespace detail {
+/**
+ * @brief Reduce node tree into column tree by aggregating each property of column.
+ *
+ * @param tree json node tree to reduce (modified in-place, but restored to original state)
+ * @param col_ids column ids of each node (modified in-place, but restored to original state)
+ * @param row_offsets row offsets of each node (modified in-place, but restored to original state)
+ * @param stream The CUDA stream to which kernels are dispatched
+ * @return A tuple containing the column tree, identifier for each column and the maximum row index
+ * in each column
+ */
+CUDF_EXPORT
+std::tuple<csr, column_tree_properties> reduce_to_column_tree(
+  tree_meta_t& tree,
+  device_span<NodeIndexT const> original_col_ids,
+  device_span<size_type const> row_offsets,
+  bool is_array_of_arrays,
+  NodeIndexT const row_array_parent_col_id,
```

Comment:
Please use `const` consistently.
```suggestion
  bool const is_array_of_arrays,
  NodeIndexT const row_array_parent_col_id,
```

Reply:
I would recommend not using `const` for parameter types unless they are pass by reference. The `const` provides no help to the caller since the pass-by-copy ensures the parameter value is not changed from the caller's perspective. It also limits the implementation which may now require a separate variable (register) to manage a value that otherwise would not be `const`.

Reply:
Thanks for sharing the insight @davidwendt . +1 on the above comment.
I do have one doubt: if the pass by value argument is const, will it help compiler to do optimizations better? (or make compiler work easier(faster) because it already knows it's const).

Reply:
There's a middle ground I've heard of: avoid const by-value in the definition (visible to user), but use const in the definition to prevent the parameter value changing somewhere in the code.
I personally don't use const here, but can understand the (mild) usefulness in the definition.

> It also limits the implementation which may now require a separate variable (register) to manage a value that otherwise would not be const.

@davidwendt could you explain this, I'm not following?

Reply:
Just that the caller does not need to know your implementation details regarding the parameter. If you want to use the parameter as non-const in your implementation that should be up to you.
Using `const` in the definition and not in the declaration seems like a hack (to me) though I know it is supported by the C++ spec.